pageserver/src/basebackup.rs

```diff
@@ -133,6 +134,84 @@ where
     ctx: &'a RequestContext,
 }
 
+/// A sink that accepts SLRU blocks ordered by key and forwards
+/// full segments to the archive.
+struct SlruSegmentsBuilder<'a, 'b, W>
+where
+    W: AsyncWrite + Send + Sync + Unpin,
+{
+    ar: &'a mut Builder<&'b mut W>,
+    buf: Vec<u8>,
+    current_segment: Option<(SlruKind, u32)>,
+}
+
+impl<'a, 'b, W> SlruSegmentsBuilder<'a, 'b, W>
+where
+    W: AsyncWrite + Send + Sync + Unpin,
+{
+    fn new(ar: &'a mut Builder<&'b mut W>) -> Self {
+        Self {
+            ar,
+            buf: Vec::new(),
+            current_segment: None,
+        }
+    }
+
+    async fn add_block(&mut self, key: &Key, block: Bytes) -> anyhow::Result<()> {
+        let (kind, segno, _) = key_to_slru_block(*key)?;
+
+        if kind == SlruKind::Clog {
+            ensure!(block.len() == BLCKSZ as usize || block.len() == BLCKSZ as usize + 8);
+        } else {
+            ensure!(block.len() == BLCKSZ as usize);
+        }
+
+        let segment = (kind, segno);
+        match self.current_segment {
+            None => {
+                self.current_segment = Some(segment);
+                self.buf
+                    .extend_from_slice(block.slice(..BLCKSZ as usize).as_ref());
+            }
+            Some(current_seg) if current_seg == segment => {
+                self.buf
+                    .extend_from_slice(block.slice(..BLCKSZ as usize).as_ref());
+            }
+            Some(_) => {
+                self.flush().await?;
+
+                self.current_segment = Some(segment);
+                self.buf
+                    .extend_from_slice(block.slice(..BLCKSZ as usize).as_ref());
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn flush(&mut self) -> anyhow::Result<()> {
+        let nblocks = self.buf.len() / BLCKSZ as usize;
+        let (kind, segno) = self.current_segment.take().unwrap();
+        let segname = format!("{}/{:>04X}", kind.to_str(), segno);
+        let header = new_tar_header(&segname, self.buf.len() as u64)?;
+        self.ar.append(&header, self.buf.as_slice()).await?;
+
+        trace!("Added to basebackup slru {} relsize {}", segname, nblocks);
+
+        self.buf.clear();
+
+        Ok(())
+    }
```

Comment:
Seems to me like instead of `extend_from_slice` (= effectively memcopying the returned page images into `self.buf`), we could make `self.buf` store the `Bytes` instead, and consume them in `self.flush()`.

That optimization saves 1 memcopy.
Since this PR adds 1 memcopy, we'd break even.

This should become a follow-up.

Reply:
We can skip the intermediate buffer by getting the underlying writer from the archive builder ([get_mut](https://docs.rs/tokio-tar/latest/tokio_tar/struct.Builder.html#method.get_mut)) and writing straight to it. We'd have to be careful to apply the same padding logic as `tokio-tar`. I'm guessing this is what you meant.

Reply:
Follow up created: https://github.com/neondatabase/neon/issues/6434

Reply:
I understand Christians idea ("we could make self.buf store the `Bytes` instead") like a `Vec<Bytes>` which I was thinking of proposing, until I checked that tokio-tar requires a `tokio::io::Read` and will not benefit from a `BufRead`. Looking around a bit, we would only need to turn the `Vec<Bytes>` into a stream with `futures::stream::from_iter`, then use `tokio_util::io::StreamReader` to expose that as `tokio::io::Read`.